### PR TITLE
[477 merge] CP-27391: datamodel option type support

### PR DESCRIPTION
**Pull request by edwintorok**
This is @kc284 's work on supporting the datamodel option type in the SDK.
This has to be merged together with the corresponding PR in XAPI:
https://github.com/xapi-project/xen-api/pull/3544

See internal jenkins build on feature/REQ477/probe builds 66 and 72 that have successfully built the rpms and the SDK.

### CODE REVIEW

File: csharp/gen_csharp_binding.ml [hunk 1]
```diff
@@ -1444,27 +1451,42 @@ and simple_convert_from_proxy thing ty =
       (exposed_class_name name) thing
   | Set(Int)            ->
     sprintf "Helper.StringArrayToLongArray(%s)" thing
-  | _                   -> assert false
+  | _ as x              -> eprintf "%s" (Types.to_string x); assert false
```
Comment (lindig) [inline csharp/gen_csharp_binding.ml]: This is equivalent to
```
| x -> ...
```

File: csharp/gen_csharp_binding.ml [hunk 2]
```diff
@@ -1407,13 +1413,14 @@ and convert_from_hashtable fname ty =
...
-    sprintf "new %s((Proxy_%s)table[%s])"
+    sprintf "(%s)Marshalling.convertStruct(typeof(%s), Marshalling.ParseHashTable(table, %s));"
       (exposed_class_name name) (exposed_class_name name) field
   | Set(Record name)    ->
     sprintf "%s.ProxyArrayToObjectList(Marshalling.ParseStringArray(%s))"
       (exposed_class_name name) field
   | Set(Int)            -> sprintf "Marshalling.ParseLongArray(table, %s)" field
-  | _                   -> assert false
+  | Option x            -> convert_from_hashtable fname x
+  | _ as x              -> eprintf "%s %s" fname (Types.to_string x); assert false
```
Comment (lindig) [inline csharp/gen_csharp_binding.ml]:
```
| x -> ...
```

File: csharp/gen_csharp_binding.ml [hunk 3]
```diff
@@ -1348,7 +1353,12 @@ and exposed_type = function
                                  (exposed_type v)
   | Record name             -> exposed_class_name name
   | Set(Record name)        -> sprintf "List<%s>" (exposed_class_name name)
-  | _                       -> assert false
+  | Option Int              -> "long?"
+  | Option Float            -> "double?"
+  | Option Bool             -> "bool?"
+  | Option DateTime         -> "DateTime?"
+  | Option x                -> exposed_type x
+  | _ as x                  -> eprintf "%s" (Types.to_string x); assert false
```
Comment (lindig) [inline csharp/gen_csharp_binding.ml]:
```
| x ->
```

File: csharp/gen_csharp_binding.ml [hunk 4]
```diff
@@ -1326,6 +1326,11 @@ and proxy_type = function
   | Enum _              -> "string"
   | Map _               -> "Object"
   | Record name         -> "Proxy_" ^ exposed_class_name name
+  | Option Float        -> "double?"
+  | Option Bool         -> "bool?"
+  | Option DateTime     -> "DateTime?"
+  | Option x            -> proxy_type x
+  | _ as x              -> eprintf "%s" (Types.to_string x); assert false
```
Comment (lindig) [inline csharp/gen_csharp_binding.ml]:
```
| x ->
```

File: c/gen_c_binding.ml [hunk 5]
```diff
@@ -222,6 +223,13 @@ and write_decl {name=classname; contents=contents; description=description;
   let full_stop = if Astring.String.is_suffix ~affix:"." description then "" else "."
   in
 
+  let rec get_needed x =
+  match x with
+    | Field fr -> find_needed'' needed fr.ty
```
Comment (minishrink) [inline c/gen_c_binding.ml]: Minor suggestion: could just write `let rec get_needed = function`?

File: c/gen_c_binding.ml [hunk 6]
```diff
@@ -342,8 +350,7 @@ and record_fields contents needed =
 and record_field needed prefix content =
   match content with
```
Comment (minishrink) [inline c/gen_c_binding.ml]: Could also do that here?

**Comment by edwintorok** [comment]
Squashed the fixups

**Comment by edwintorok** [comment]
assert false is for code that should never be reached, its a bug in the code if it is. Using GADTs would probably reduce the number of places that is needed, but its not always easy to refactor code to  use GADTs.

**Comment by edwintorok** [comment]
Agreed on cleaning up the match to use function instead, there is a lot of existing code that does it that way, I think that cleanup should be part of another PR

**Pull request merged**